### PR TITLE
spi: In-memory policy store implemetation

### PR DESCRIPTION
**Pull request by Izzzu**
## What this PR changes/adds

- Add PolicyStore interface.
- Implement in memory version of PolicyStore.
- Add InMemoryPolicyStoreExtension

## Why it does that

Implementation of PolicyStore according to definitions in upstream issue: [#803](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/issues/803)

## Further notes

This PR contains only changes related to PolicyStore. Following points willbe addressed in the next internal PR:

1. 
During implementation I noticed that the findAll by QuerySpec method
```
Stream<Policy> findAll(QuerySpec spec);
```
uses code that is duplicated across a many store implementations, eg. InMemoryContractNegotiationStore, InMemoryContractDefinitionStore, CosmosContractDefinitionStore etc. which can be refactored.

2. 
Additionally, an inconsistency was found, that in some cases if sortField is invalid the _findAll_ method returns an empty stream, but in some cases it returns the whole stream.
Cosmos store implementations are following the first strategy - returning an empty stream if sortField is invalid, therefore in memory version should always be consistent with that contract.

## Linked Issue(s)
 #136 

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [x] added/updated relevant documentation?
- [x] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: extensions/in-memory/policy-store-memory/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+/*
+ *  Copyright (c) 2020, 2021 Microsoft Corporation
```
Comment (cpeeyush) [inline extensions/in-memory/policy-store-memory/build.gradle.kts:2]: nit: year 2022.

File: extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
+package org.eclipse.dataspaceconnector.policy.store.memory;
```
Comment (cpeeyush) [inline extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java:1]: nit: copyright header is missing from the newly created files.

Comment (Izzzu) [reply extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java:14]: added

File: extensions/in-memory/policy-store-memory/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+plugins {
+    `java-library`
+}
+
+dependencies {
+    api(project(":spi"))
+    implementation(project(":common:util"))
+
+
```
Comment (ouphi) [inline extensions/in-memory/policy-store-memory/build.gradle.kts:25]: nit: 2 extra spaces here and one missing space below before publishing

File: extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    @Override
+    public void save(Policy policy) {
+        lockManager.writeLock(() -> policiesById.put(policy.getUid(), policy));
+    }
+
+    @Override
+    public Policy delete(String policyId) {
+        return lockManager.writeLock(() -> Optional.ofNullable(policiesById.remove(policyId)).orElse(null));
```
Comment (ouphi) [inline extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java:64]: Why did you need to wrap it into an Optional?
I think that only:
`return lockManager.writeLock(() -> policiesById.remove(policyId));` would work the same no? As `policiesById.remove` returns null when the key does not exist.

File: extensions/in-memory/policy-store-memory/src/test/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStoreTest.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    void findById_whenPresent() {
+        Policy policy = Policy.Builder.newInstance().build();
+        store.save(policy);
+
+        assertThat(store.findById(policy.getUid())).isEqualTo(policy);
+    }
+
+    @Test
+    void findById_whenMissing() {
+        assertThat(store.findById(UUID.randomUUID().toString())).isNull();
```
Comment (ouphi) [inline extensions/in-memory/policy-store-memory/src/test/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStoreTest.java:45]: nit: From a code review I got from Paul I think that their personal preference is to use an id called something like "not-exists" for this type of tests where you query/delete a non existing item.

**Review by ouphi — APPROVED** [review]

File: extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java (hunk 4 above)
Comment (Izzzu) [reply extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java:64]: true, it's not necessary, thanks

File: extensions/in-memory/policy-store-memory/src/test/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStoreTest.java (hunk 5 above)
Comment (Izzzu) [reply extensions/in-memory/policy-store-memory/src/test/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStoreTest.java:45]: ok, thanks, will change that

File: extensions/in-memory/policy-store-memory/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+/*
+ *  Copyright (c) 2020-2022 Microsoft Corporation
```
Comment (cpeeyush) [inline extensions/in-memory/policy-store-memory/build.gradle.kts:2]: extensions/in-memory/build.gradle.kts

Comment (Izzzu) [reply extensions/in-memory/policy-store-memory/build.gradle.kts:2]: good catch, added

File: spi/contract-spi/src/main/java/org/eclipse/dataspaceconnector/spi/contract/policy/store/PolicyStore.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.util.stream.Stream;
+
+/**
+ * Persists {@link Policy}.
+ */
+public interface PolicyStore {
+
+    /**
+     * Findss the policy by id.
```
Comment (cpeeyush) [inline spi/contract-spi/src/main/java/org/eclipse/dataspaceconnector/spi/contract/policy/store/PolicyStore.java:16]:
```suggestion
     * Finds the policy by id.
```

File: spi/contract-spi/src/main/java/org/eclipse/dataspaceconnector/spi/contract/policy/store/PolicyStore.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+     *
+     * @param policyId id of the policy.
+     * @return found Policy or null if not found.
```
Comment (cpeeyush) [inline spi/contract-spi/src/main/java/org/eclipse/dataspaceconnector/spi/contract/policy/store/PolicyStore.java:19]:
```suggestion
     * @return {@link Policy} or null if not found.
```

File: spi/contract-spi/src/main/java/org/eclipse/dataspaceconnector/spi/contract/policy/store/PolicyStore.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+     * @throws EdcPersistenceException if something goes wrong.
+     */
+    @Nullable
+    Policy findById(String policyId);
+
+    /**
+     * Find stream of policies in the store based on query spec.
+     *
+     * @param spec query specification.
+     * @return stream of found policies.
```
Comment (cpeeyush) [inline spi/contract-spi/src/main/java/org/eclipse/dataspaceconnector/spi/contract/policy/store/PolicyStore.java:29]:
```suggestion
     * @return A {@link Stream} of {@link Policy}. Might be empty, never null.
```

File: spi/contract-spi/src/main/java/org/eclipse/dataspaceconnector/spi/contract/policy/store/PolicyStore.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+     * @param policy to be saved.
+     * @throws EdcPersistenceException if something goes wrong.
+     */
+    void save(Policy policy);
+
+    /**
+     * Deletes a policy for the given id.
+     *
+     * @param policyId id of the policy to be removed.
+     * @return deleted policy or null if policy not found.
```
Comment (cpeeyush) [inline spi/contract-spi/src/main/java/org/eclipse/dataspaceconnector/spi/contract/policy/store/PolicyStore.java:46]:
```suggestion
     * @return Deleted {@link Policy} or null if policy not found.
```

File: extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+        //filter
+        var andPredicate = spec.getFilterExpression().stream().map(this::toPredicate).reduce(x -> true, Predicate::and);
+        Stream<T> stream  = streamOfAll.filter(andPredicate);
+
+        //sort
+        var sortField = spec.getSortField();
+
+        if (sortField != null) {
+            // if the sortfield doesn't exist on the object -> return empty
```
Comment (cpeeyush) [inline extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java:77]:
```suggestion
            // if the sort field doesn't exist on the object -> return empty
```

File: extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+import static org.eclipse.dataspaceconnector.common.reflection.ReflectionUtil.propertyComparator;
+
+/**
+ * An in-memory, threadsafe policy store.
+ * This implementation is intended for testing purposes only.
+ */
+public class InMemoryPolicyStore implements PolicyStore {
+
+    private final LockManager lockManager = new LockManager(new ReentrantReadWriteLock());
```
Comment (cpeeyush) [inline extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java:41]: Can we inject it via constructor ? So it can be mocked in UTs.

File: extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import static org.eclipse.dataspaceconnector.common.reflection.ReflectionUtil.propertyComparator;
+
+/**
+ * An in-memory, threadsafe policy store.
+ * This implementation is intended for testing purposes only.
+ */
+public class InMemoryPolicyStore implements PolicyStore {
```
Comment (cpeeyush) [inline extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java:39]: As per the upstream 803 issue. EdcPersistenceException is thrown when something goes wrong. Looks like that is not covered here and also UTs are missing for that.

File: extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    public void save(Policy policy) {
+        lockManager.writeLock(() -> policiesById.put(policy.getUid(), policy));
+    }
+
+    @Override
+    public Policy delete(String policyId) {
+        return lockManager.writeLock(() -> policiesById.remove(policyId));
+    }
+
+    private <T> Stream<T> applyQuery(QuerySpec spec, Stream<T> streamOfAll, Class<?> clazz) {
```
Comment (cpeeyush) [inline extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java:67]: nit: Looks like this method is not going to be reused by somewhere else. Here this will always return Stream of Policies so it can be a bit simplified by not using generics here.

But it is more of a coding style so you can decide.

**Comment by cpeeyush** [comment]
Added few comments.

File: extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java (hunk 14 above)
Comment (Izzzu) [reply extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java:67]: indeed, I can simplify it here. It's a left over after splitting the code into 2 PRs - this one and the next PR that introduces the refactoring. In the next Pr this method will be extracted to a class. But to avoid confusion I'll simplify it here. Thanks

File: extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java (hunk 13 above)
Comment (cpeeyush) [reply extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java:39]: I see your point, maybe then we do not need to do this here we can keep it simple as other stores.

File: extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+
+    private final LockManager lockManager;
+    private final Map<String, Policy> policiesById =  new HashMap<>();
+
+    public InMemoryPolicyStore(LockManager lockManager) {
+        this.lockManager = lockManager;
+    }
+
+    public InMemoryPolicyStore() {
+        this.lockManager = new LockManager(new ReentrantReadWriteLock());
```
Comment (cpeeyush) [inline extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java:49]: nit: lock with fairness. 
```suggestion
        this.lockManager = new LockManager(new ReentrantReadWriteLock(true));
```

Comment (Izzzu) [reply extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java]: changed
